### PR TITLE
feature: add SingleLineEmptyBodyFixer

### PR DESCRIPTION
**Pull request by kubawerlos**
Closes https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/6932
Closes https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/5475
Replaces https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/5757

### CODE REVIEW

File: src/Fixer/Basic/EmptyBodyFixer.php [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace PhpCsFixer\Fixer\Basic;
+
+use PhpCsFixer\AbstractFixer;
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
+use PhpCsFixer\Tokenizer\Tokens;
+
+final class EmptyBodyFixer extends AbstractFixer
```
Comment (Wirone) [inline src/Fixer/Basic/EmptyBodyFixer.php:23]: I don't like `empty_body` name. @keradus once said fixers should be named from desired state (after fix), which here is "single-line empty body", not "empty body" itself. It could be just `empty_body` if it also handled comments within braces (extract them outside of the body), which is not the case by looking at the tests.

File: src/Fixer/Basic/EmptyBodyFixer.php [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    protected function applyFix(\SplFileInfo $file, Tokens $tokens): void
+    {
+        for ($index = $tokens->count() - 1; $index > 0; --$index) {
+            if (!$tokens[$index]->isGivenKind([T_CLASS, T_FUNCTION])) {
+                continue;
+            }
+
+            $openBraceIndex = $tokens->getNextTokenOfKind($index, ['{', ';']);
+            \assert(\is_int($openBraceIndex));
```
Comment (Wirone) [inline src/Fixer/Basic/EmptyBodyFixer.php:61]: Is it project's convention to use `assert()` in the fixers? I didn't see it before.

Comment (kubawerlos) [reply src/Fixer/Basic/EmptyBodyFixer.php:61]: Is it project's convention to NOT use `assert()` in the fixers? If so, I'd like to challenge that as IMHO it is much better than PHPDoc (enforces type being what it claims to be, not only promises that).

File: src/Fixer/Basic/EmptyBodyFixer.php (hunk 1 above)
Comment (kubawerlos) [reply src/Fixer/Basic/EmptyBodyFixer.php:23]: "single-line empty body" would also be true for a body like `{ return true; }` or `{ /* @todo: implement this */ }`.

By empty body, I meant literally empty, like nothing between `{` and `}`. Would `abbreviated_as_pair_of_braces_and_placed_on_the_same_line_as_the_previous_symbol_and_separated_by_a_space` be a better name? 😉

Comment (Wirone) [reply src/Fixer/Basic/EmptyBodyFixer.php:23]: > "single-line empty body" would also be true for a body like { return true; } or { /* @todo: implement this */ }.

No, because those bodies are not _empty_ 😉.

On the other hand, you can have:

```php
class Foo {
    public function __construct()
    {




    }
}
```

and the body _is_ empty (so it fits fixer's current name), but is _not_ single line. So if desired state is single-line empty body (`{}`), it should be named `single_line_empty_body`.

File: src/Fixer/Basic/EmptyBodyFixer.php (hunk 2 above)
Comment (Wirone) [reply src/Fixer/Basic/EmptyBodyFixer.php:61]: It looks like some ADR to be made, I wouldn't introduce such new approach in PR with new fixer, but rather use [RFC](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/discussions/categories/rfc) discussion for that 🙂. To be honest I have never used native `assert()` and I don't even know how it would behave here (from user perspective in CLI). I am 100% for such improvements, but I think it's not good moment to introduce it 😉.

**Review by Wirone — APPROVED** [review]
Do we want add this fixer to `PhpCsFixer` set? 🙂 It can be done here, or later too - 👍 from me anyway. Great job @kubawerlos 👏

**Comment by kubawerlos** [comment]
(missclick 🤦🏼 )

**Comment by kubawerlos** [comment]
> Do we want add this fixer to `PhpCsFixer` set? slightly_smiling_face It can be done here, or later too - +1 from me anyway. Great job @kubawerlos clap

Let's leave @keradus this do decide and do if he likes it.

File: tests/Fixer/Basic/SingleLineEmptyBodyFixerTest.php
Comment (Wirone) [inline tests/Fixer/Basic/SingleLineEmptyBodyFixerTest.php:1]: Last thought: can we add test for PHP8 with constructor using property promotion and empty body?

**Review by Wirone — CHANGES_REQUESTED** [review]

**Review by Wirone — APPROVED** [review]

**Comment by Wirone** [comment]
Thank you very much @kubawerlos 🍻

**Comment by deleugpn** [comment]
has this fixer been released?

File: doc/rules/basic/single_line_empty_body.rst [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+--------
+
+Example #1
+~~~~~~~~~~
+
+.. code-block:: diff
+
+   --- Original
+   +++ New
+    <?php function foo(
```
Comment (mvorisek) [inline doc/rules/basic/single_line_empty_body.rst:18]: PSR-12 standard requires:
- for non-anonymous function/class: line break before `{` and one before `}`, this is common coding standard and it helps to reduce diff noise
- for anonymus function/class no line break

So the rule should be configureable for these usecases.

Comment (Wirone) [reply doc/rules/basic/single_line_empty_body.rst:18]: PSR-12 is replaced by [PER-CS](https://www.php-fig.org/per/coding-style/) which enforces/encourages `{}` for both scenarios, so there's no point in supporting outdated coding standard.

**Comment by romm** [comment]
A bit late, just passing by to say thank you for this feature I've been waiting for a long time! ❤️

**Comment by raveren** [comment]
To anyone coming to this issue from google, here's how you use this rule:

Just add the rule `'single_line_empty_body'` to your config, example:

```php
// as close to PER as possible 
return (new PhpCsFixer\Config())->setRules([
    '@PSR12'                 => true,
    'function_declaration'   => ['closure_function_spacing' => 'none', 'closure_fn_spacing' => 'none'],
    'no_break_comment'       => false,
    'single_line_empty_body' => true, // this!
])
    ->setFinder($finder)
```
